crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopFixpoint.hs

```diff
@@ -0,0 +1,748 @@
+import qualified Lang.Crucible.Simulator.RegMap as C
+import qualified Lang.Crucible.Simulator as C
+
+import qualified Lang.Crucible.LLVM.Bytes as C
+import qualified Lang.Crucible.LLVM.DataLayout as C
+import qualified Lang.Crucible.LLVM.MemModel as C
+import qualified Lang.Crucible.LLVM.MemModel.MemLog as C hiding (Mem)
+import qualified Lang.Crucible.LLVM.MemModel.Pointer as C
+import qualified Lang.Crucible.LLVM.MemModel.Type as C
+
+
+data FixpointStatus = BeforeFixpoint | ComputeFixpoint | CheckFixpoint | AfterFixpoint
+  deriving (Eq, Ord, Show)
+
+data FixpointEntry sym tp = FixpointEntry
+  { headerValue :: W4.SymExpr sym tp
+  , bodyValue :: W4.SymExpr sym tp
+  }
+
+instance OrdF (W4.SymExpr sym) => OrdF (FixpointEntry sym) where
+  compareF x y = case compareF (headerValue x) (headerValue y) of
+    LTF -> LTF
+    EQF -> compareF (bodyValue x) (bodyValue y)
+    GTF -> GTF
+
+instance OrdF (FixpointEntry sym) => W4.TestEquality (FixpointEntry sym) where
+  testEquality x y = case compareF x y of
+    EQF -> Just Refl
+    _ -> Nothing
+
+instance W4.TestEquality (FixpointEntry sym) => C.EqF (FixpointEntry sym) where
+  eqF x y = isJust $ W4.testEquality x y
+
+data MemFixpointEntry sym = forall w . (1 <= w) => MemFixpointEntry
+  { memFixpointEntrySym :: sym
+  , memFixpointEntryJoinVariable :: W4.SymBV sym w
+  }
+
+instance W4.TestEquality (W4.SymExpr sym) => Eq (MemFixpointEntry sym) where
+  (MemFixpointEntry _ x_join_variable) == (MemFixpointEntry _ y_join_variable)
+    | Just Refl <- W4.testEquality x_join_variable y_join_variable =
+      True
+    | otherwise = False
+
+data FixpointStuff sym blocks = forall args . FixpointStuff
+  { fixpointStatus :: FixpointStatus
+  , fixpointBlockId :: C.Some (C.BlockID blocks)
+  , fixpointAssumptionFrameIdentifier :: C.FrameIdentifier
+  , fixpointSubstitution :: MapF (W4.SymExpr sym) (FixpointEntry sym)
+  , fixpointRegMap :: C.RegMap sym args
+  , fixpointMemSubstitution :: Map (Natural, Natural, Natural) (MemFixpointEntry sym, C.StorageType)
+  , fixpointLoopCondition :: W4.Pred sym
+  , fixpointLoopIndexBound :: LoopIndexBound sym
+  }
+
+type FixpointMonad sym = StateT (MapF (W4.SymExpr sym) (FixpointEntry sym)) IO
+
+
+joinRegEntries ::
+  (?logMessage :: String -> IO (), C.IsSymInterface sym) =>
+  sym ->
+  Ctx.Assignment (C.RegEntry sym) ctx ->
+  Ctx.Assignment (C.RegEntry sym) ctx ->
+  FixpointMonad sym (Ctx.Assignment (C.RegEntry sym) ctx)
+joinRegEntries sym = Ctx.zipWithM (joinRegEntry sym)
+
+joinRegEntry ::
+  (?logMessage :: String -> IO (), C.IsSymInterface sym) =>
+  sym ->
+  C.RegEntry sym tp ->
+  C.RegEntry sym tp ->
+  FixpointMonad sym (C.RegEntry sym tp)
+joinRegEntry sym left right = case C.regType left of
+  C.LLVMPointerRepr w
+    | List.isPrefixOf "cmacaw_reg" (show $ W4.printSymNat $ C.llvmPointerBlock (C.regValue left))
+    , List.isPrefixOf "cmacaw_reg" (show $ W4.printSymExpr $ C.llvmPointerOffset (C.regValue left)) -> do
```

Comment:
This looks a bit fragile and scary. Also quite macaw-specific... is there some more structured way we could accomplish this that is resilient to breaking?

Reply:
I agree, this is a convenient way of checking that a register is locally used inside the loop, but doesn't need to be part of the fixpoint. It can be improved later